modules/table/src/main/java/org/apache/ignite/internal/utils/RebalanceUtil.java

```diff
@@ -0,0 +1,135 @@
+ */
+public class RebalanceUtil {
+
+    /**
+     * Update keys that related to rebalance algorithm in Meta Storage. Keys are specific for partition.
+     *
+     * @param partId Unique identifier of a partition.
+     * @param baselineNodes Nodes in baseline.
+     * @param partitions Number of partitions in a table.
+     * @param replicas Number of replicas for a table.
+     * @param revision Revision of Meta Storage that is specific for the assignment update.
+     * @param metaStorageMgr Meta Storage manager.
+     */
+    public static void updateAssignmentsKeys(String partId, Collection<ClusterNode> baselineNodes, int partitions, int replicas,
+            long revision, MetaStorageManager metaStorageMgr) {
+        ByteArray partChangeTriggerKey = partChangeTriggerKey(partId);
+
+        ByteArray partAssignmentsPendingKey = partAssignmentsPendingKey(partId);
+
+        ByteArray partAssignmentsPlannedKey = partAssignmentsPlannedKey(partId);
+
+        ByteArray partAssignmentsStableKey = partAssignmentsStableKey(partId);
+
+        byte[] partAssignmentsBytes = ByteUtils.toBytes(AffinityUtils.calculateAssignments(baselineNodes, partitions, replicas));
+
+        //    if empty(partition.change.trigger.revision) || partition.change.trigger.revision < event.revision:
+        //        if empty(partition.assignments.pending) && partition.assignments.stable != calcPartAssighments():
+        //            partition.assignments.pending = calcPartAssignments()
+        //            partition.change.trigger.revision = event.revision
+        //        else:
+        //            if partition.assignments.pending != calcPartAssignments
+        //                partition.assignments.planned = calcPartAssignments()
+        //                partition.change.trigger.revision = event.revision
+        //            else
+        //                remove(partition.assignments.planned)
+        //    else:
+        //        skip
+        var iif = If.iif(or(exists(partChangeTriggerKey), value(partChangeTriggerKey).lt(ByteUtils.longToBytes(revision))),
+                If.iif(and(exists(partAssignmentsPendingKey), value(partAssignmentsStableKey).ne(partAssignmentsBytes)),
+                        ops(
+                                put(partAssignmentsPendingKey, partAssignmentsBytes),
+                                put(partChangeTriggerKey, ByteUtils.longToBytes(revision))
+                        ).yield(),
+                        If.iif(value(partAssignmentsPendingKey).ne(partAssignmentsBytes),
+                                ops(
+                                        put(partAssignmentsPlannedKey, partAssignmentsBytes),
+                                        put(partChangeTriggerKey, ByteUtils.longToBytes(revision))
+                                ).yield(),
+                                ops(remove(partAssignmentsPlannedKey)).yield())),
+                ops().yield());
+
+        metaStorageMgr.invoke(iif);
```

Comment:
This future must be returned as a result for further handling of errors, I believe.

Reply:
done